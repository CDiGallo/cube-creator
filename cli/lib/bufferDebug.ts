import stream from 'readable-stream'
import type Pipeline from 'barnard59-core/lib/Pipeline'
import type { BoundBaseObserver } from '@opentelemetry/api-metrics'
import { bufferObserver } from './otel/metrics'

const { finished } = stream

function bufferStatePair({ state, step }: any): { key: string; value: number } {
  const key = step.ptr.value // `[${index}] (${mode}) ${step.ptr.value} (${state.length}/${state.highWaterMark})`
  const value = state.length > 0 ? Math.round(state.length / state.highWaterMark * 100.0) : 0

  return { key, value }
}

function bufferInfo(pipeline: Pipeline) {
  const steps = pipeline.children

  if (!steps) {
    return
  }

  return steps.reduce<Record<string, number>>((data, step, index) => {
    if ('_writableState' in step.stream) {
      const { key, value } = bufferStatePair({
        index,
        mode: 'write',
        state: step.stream._writableState,
        step,
      })

      data[key] = value
    }

    if ('_readableState' in step.stream) {
      const { key, value } = bufferStatePair({
        index,
        mode: 'read',
        state: step.stream._readableState,
        step,
      })

      data[key] = value
    }

    return data
  }, {})
}

function bufferDebug(pipeline: Pipeline, jobUri: string, { interval = 10 } = {}): void {
  let done = false

  finished(pipeline.stream, () => {
    done = true
  })

  const pid = process.pid.toString()
  const boundMeters = new Map<string, BoundBaseObserver>()
  function getMeter(step: string) {
    const bound = boundMeters.get(step) || bufferObserver.bind({
      jobUri,
      pipeline: pipeline.ptr.value,
      step,
      pid,
    })

    boundMeters.set(step, bound)

    return bound
  }

  const next = async () => {
    if (done) {
      return
    }

    const data = bufferInfo(pipeline)

    if (data) {
      [...Object.entries(data)].forEach(([step, value]) => {
        getMeter(step).update(value)
      })
    }

    setTimeout(next, interval)
  }

  next()
}

export default bufferDebug
