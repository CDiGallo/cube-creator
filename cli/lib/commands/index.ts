export { default as transform } from './transform'
export { default as publish } from './publish'
export { default as importCube } from './import'
