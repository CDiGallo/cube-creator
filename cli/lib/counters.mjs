import { PassThrough } from 'readable-stream'
import { quadCounter } from './otel/metrics.mjs'

export function quads(name) {
  const bound = quadCounter.bind({
    name,
  })
  const forwarder = new PassThrough({
    objectMode: true,
  })

  forwarder.on('data', () => {
    bound.add(1)
  })

  return forwarder
}
