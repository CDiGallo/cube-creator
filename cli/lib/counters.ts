import { Quad } from 'rdf-js'
import type { Context } from 'barnard59-core/lib/Pipeline'
import { PassThrough } from 'readable-stream'
import through2 from 'through2'
import { csvw } from '@tpluscode/rdf-ns-builders/strict'
import { quadCounter } from './otel/metrics'

export function quads(this: Context, name: string) {
  const jobUri = this.variables.get('jobUri')
  const forwarder = new PassThrough({
    objectMode: true,
  })

  forwarder.on('data', () => {
    quadCounter.add(1, {
      name,
      job_id: jobUri.substr(jobUri.lastIndexOf('/') + 1),
    })
  })

  return forwarder
}

export function observeCsvwMetadata(this: Context) {
  const lastTransformed = this.variables.get('lastTransformed')

  return through2.obj(function (chunk: Quad, enc, callback) {
    if (chunk.predicate.equals(csvw.rownum)) {
      lastTransformed.row = Number.parseInt(chunk.object.value, 10)
    }

    this.push(chunk)
    callback()
  })
}
