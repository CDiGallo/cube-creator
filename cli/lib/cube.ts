import rdf from 'rdf-ext'
import { DatasetCore } from 'rdf-js'
import { cc } from '@cube-creator/core/namespace'
import { GraphPointer } from 'clownface'

export const getObservationSetId = ({ dataset }: { dataset: DatasetCore }) => {
  const cubeId = [...dataset.match(null, cc.cube)][0].object.value

  if (cubeId.endsWith('/') || cubeId.endsWith('#')) {
    return rdf.namedNode(`${cubeId}observation/`)
  }

  return rdf.namedNode(`${cubeId}/observation/`)
}

export function getCubeId({ ptr }: { ptr: GraphPointer }) {
  return ptr.out(cc.cube).term || ''
}
