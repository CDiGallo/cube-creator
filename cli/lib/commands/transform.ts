import path from 'path'
import { NamedNode } from 'rdf-js'
import $rdf from 'rdf-ext'
import { fileToDataset } from 'barnard59'
import { Debugger } from 'debug'
import { setupAuthentication } from '../auth'
import Runner from 'barnard59/lib/runner'
import bufferDebug from 'barnard59/lib/bufferDebug'
import clownface from 'clownface'
import namespace from '@rdfjs/namespace'
import { rdf } from '@tpluscode/rdf-ns-builders'

const ns = {
  pipelines: namespace('https://pipeline.described.at/'),
}

interface RunOptions {
  debug: boolean
  to: string
  project: string
  variable: Map<string, string>
  enableBufferMonitor: boolean
  authParam: Map<string, string>
}

export default function (pipelineId: NamedNode, log: Debugger) {
  const basePath = path.resolve(__dirname, '../../')

  return async function (command: RunOptions) {
    const { to, project, debug, enableBufferMonitor, variable } = command

    log.enabled = debug

    const authConfig = {
      params: command.authParam,
    }
    setupAuthentication(authConfig, log)

    const pipelinePath = (filename: string) => path.join(basePath, `./pipelines/${filename}.ttl`)
    const dataset = $rdf.dataset()
      .merge(await fileToDataset('text/turtle', pipelinePath('main')))
      .merge(await fileToDataset('text/turtle', pipelinePath('from-api')))
      .merge(await fileToDataset('text/turtle', pipelinePath(`to-${to}`)))

    log('Running project %s', project)
    const pipeline = clownface({ dataset }).namedNode(pipelineId)
    variable.set('projectUri', project)

    pipeline.addOut(ns.pipelines.variables, set => {
      variable.forEach((value, key) => {
        set.addOut(ns.pipelines.variable, variable => {
          variable.addOut(rdf.type, ns.pipelines.Variable)
            .addOut(ns.pipelines('name'), key)
            .addOut(ns.pipelines.value, value)
        })
      })
    })

    const run = Runner.create({
      basePath: path.resolve(basePath, 'pipelines'),
      outputStream: process.stdout,
      term: pipelineId.value,
      dataset,
    })

    Runner.log.enabled = debug

    if (enableBufferMonitor) {
      bufferDebug(run.pipeline)
    }

    return run.promise
  }
}
