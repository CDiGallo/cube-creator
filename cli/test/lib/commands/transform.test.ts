import env from '@cube-creator/core/env'
import { before, describe, it } from 'mocha'
import { expect } from 'chai'
import $rdf from 'rdf-ext'
import { CONSTRUCT } from '@tpluscode/sparql-builder'
import debug from 'debug'
import { Hydra } from 'alcaeus/node'
import { rdfs, schema, sh, xsd } from '@tpluscode/rdf-ns-builders'
import transform from '../../../lib/commands/transform'
import { setupEnv } from '../../support/env'
import { client, insertTestData } from '@cube-creator/testing/lib'

describe('lib/commands/transform', function () {
  this.timeout(200000)

  const executionUrl = 'http://example.com/transformation-test'

  before(async () => {
    setupEnv()
    await insertTestData('fuseki/sample-ubd.trig')
  })

  it('produces triples', async function () {
    // given
    const runner = transform($rdf.namedNode('urn:pipeline:cube-creator#Main'), debug('test'))

    // when
    await runner({
      to: 'graph-store',
      debug: true,
      job: `${env.API_CORE_BASE}project/ubd/csv-mapping/jobs/test-job`,
      executionUrl,
    })
    await new Promise((resolve) =>
      setTimeout(resolve, 100),
    )

    // then
    const expectedGraph = $rdf.namedNode(`${env.API_CORE_BASE}cube-project/ubd/cube-data`)
    const dataset = await $rdf.dataset().import(await CONSTRUCT`?s ?p ?o`
      .FROM(expectedGraph)
      .WHERE`?s ?p ?o`
      .execute(client.query))

    expect(dataset.toCanonical()).to.matchSnapshot(this)
  })

  it('updates job when pipeline ends successfully', async function () {
    // given
    const jobUri = `${env.API_CORE_BASE}project/ubd/csv-mapping/jobs/test-job`
    const runner = transform($rdf.namedNode('urn:pipeline:cube-creator#Main'), debug('test'))

    // when
    await runner({
      to: 'graph-store',
      debug: true,
      job: jobUri,
      executionUrl,
    })

    // then
    const job = await Hydra.loadResource(jobUri)
    expect(job.representation?.root).to.matchShape({
      property: [{
        path: schema.actionStatus,
        hasValue: schema.CompletedActionStatus,
        minCount: 1,
        maxCount: 1,
      }, {
        path: rdfs.seeAlso,
        minCount: 1,
        nodeKind: sh.IRI,
      }],
    })
  })

  it('updates job when pipeline has errors', async function () {
    // given
    const jobUri = `${env.API_CORE_BASE}project/ubd/csv-mapping/jobs/broken-job`
    const runner = transform($rdf.namedNode('urn:pipeline:cube-creator#Main'), debug('test'))

    // when
    await runner({
      to: 'graph-store',
      debug: true,
      job: jobUri,
      executionUrl,
    })
    await new Promise((resolve) =>
      setTimeout(resolve, 100),
    )

    // then
    const job = await Hydra.loadResource(jobUri)
    expect(job.representation?.root).to.matchShape({
      property: [{
        path: schema.actionStatus,
        hasValue: schema.FailedActionStatus,
        minCount: 1,
        maxCount: 1,
      }, {
        path: rdfs.seeAlso,
        minCount: 1,
        nodeKind: sh.IRI,
      }, {
        path: schema.error,
        node: {
          property: {
            path: schema.description,
            datatype: xsd.string,
            minCount: 1,
          },
        },
        minCount: 1,
      }],
    })
  })
})
