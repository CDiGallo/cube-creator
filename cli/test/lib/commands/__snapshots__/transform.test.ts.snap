exports["lib/commands/transform produces triples"] = "<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1984-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1984-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"31.9\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1984-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1984\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1984-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1984-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1985-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1985-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"40.2\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1985-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1985\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1985-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1985-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1986-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1986-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"33.6\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1986-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1986\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1986-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1986-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1987-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1987-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"33\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1987-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1987\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1987-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1987-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1988-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1988-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"17.4\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1988-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1988\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1988-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1988-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1989-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1989-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"18.5\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1989-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1989\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1989-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1989-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1990-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1990-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"14.1\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1990-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1990\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1990-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1990-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1991-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1991-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"13.6\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1991-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1991\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1991-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1991-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1992-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1992-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"11\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1992-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1992\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1992-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1992-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1993-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1993-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"9.2\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1993-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1993\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1993-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1993-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1994-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1994-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"6.7\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1994-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1994\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1994-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1994-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1995-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1995-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"7.2\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1995-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1995\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1995-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1995-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1996-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1996-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"7.2\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1996-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1996\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1996-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1996-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1997-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1997-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"7.3\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1997-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1997\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1997-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1997-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1998-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1998-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"7.1\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1998-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1998\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1998-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1998-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1999-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1999-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"5.4\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1999-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"1999\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1999-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-1999-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2000-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2000-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"4.7\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2000-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"2000\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2000-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2000-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2001-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2001-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"4.4\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2001-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"2001\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2001-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2001-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2002-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2002-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"6.1\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2002-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"2002\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2002-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2002-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2003-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/aggregation> <https://environment.ld.admin.ch/foen/ubd/28/aggregation/annualmean> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2003-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/value> \"5.4\"^^<http://www.w3.org/2001/XMLSchema#float> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2003-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/dimension/year> \"2003\"^^<http://www.w3.org/2001/XMLSchema#gYear> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2003-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/station> <https://environment.ld.admin.ch/foen/ubd/28/station/blBAS> .\n<https://environment.ld.admin.ch/foen/ubd/28/so2/observation/blBAS-2003-annualmean> <https://environment.ld.admin.ch/foen/ubd/28/unit-id> \"µg/m3\" .\n";

